node/impl/full/multisig.go

```diff
@@ -0,0 +1,282 @@
+		GasLimit: 100000,
+		GasPrice: types.NewInt(1),
+	}
+
+	smsg, err := a.MpoolAPI.MpoolPushMessage(ctx, msg)
+	if err != nil {
+		return api.UndefMsigProposalOutput, nil
+	}
+
+	fmt.Println("send proposal in message: ", smsg.Cid())
+
+	wait, err := a.StateAPI.StateWaitMsg(ctx, smsg.Cid())
+	if err != nil {
+		return api.UndefMsigProposalOutput, nil
+	}
+
+	if wait.Receipt.ExitCode != 0 {
+		return api.UndefMsigProposalOutput, fmt.Errorf("proposal returned exit %d", wait.Receipt.ExitCode)
+	}
+
+	_, v, err := cbg.CborReadHeader(bytes.NewReader(wait.Receipt.Return))
+	if err != nil {
+		return api.UndefMsigProposalOutput, nil
+	}
+
+	return api.MsigProposeOutput{
+		MessageCid:    smsg.Cid(),
+		TransactionID: v,
+	}, nil
+}
+
+func (a *MsigAPI) MsigApprove(ctx context.Context, msig address.Address, txID uint64, proposer address.Address, to address.Address, amt types.BigInt, src address.Address, method uint64, params []byte) (cid.Cid, error) {
+	return a.MsigApproveOrCancel(ctx, api.MsigApprove, msig, txID, proposer, to, amt, src, method, params)
+}
+
+func (a *MsigAPI) MsigCancel(ctx context.Context, msig address.Address, txID uint64, proposer address.Address, to address.Address, amt types.BigInt, src address.Address, method uint64, params []byte) (cid.Cid, error) {
+	return a.MsigApproveOrCancel(ctx, api.MsigCancel, msig, txID, proposer, to, amt, src, method, params)
+}
+
+func (a *MsigAPI) MsigApproveOrCancel(ctx context.Context, operation api.MsigProposeResponse, msig address.Address, txID uint64, proposer address.Address, to address.Address, amt types.BigInt, src address.Address, method uint64, params []byte) (cid.Cid, error) {
+	if msig == address.Undef {
+		return cid.Undef, xerrors.Errorf("must provide multisig address")
+	}
+
+	if to == address.Undef {
+		return cid.Undef, xerrors.Errorf("must provide proposed target address")
+	}
+
+	if !types.BigPos(amt) {
+		return cid.Undef, xerrors.Errorf("must provide the positive amount that was proposed")
+	}
+
+	if src == address.Undef {
+		defaddr, err := a.WalletAPI.WalletDefaultAddress(ctx)
+		if err != nil {
+			return cid.Undef, err
+		}
+		src = defaddr
+	}
+
+	if proposer.Protocol() != address.ID {
+		proposerID, err := a.StateAPI.StateLookupID(ctx, proposer, types.EmptyTSK)
+		if err != nil {
+			return cid.Undef, err
+		}
+		proposer = proposerID
+	}
+
+	p := samsig.ProposalHashData{
+		Requester: proposer,
+		To:        to,
+		Value:     amt,
+		Method:    abi.MethodNum(method),
+		Params:    params,
+	}
+
+	pser, err := p.Serialize()
+	if err != nil {
+		return cid.Undef, err
+	}
+	phash := blake2b.Sum256(pser)
+
+	enc, err := actors.SerializeParams(&samsig.TxnIDParams{
+		ID:           samsig.TxnID(txID),
+		ProposalHash: phash[:],
+	})
+
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	var msigResponseMethod abi.MethodNum
+
+	/*
+		We pass in a MsigProposeResponse instead of MethodNum to
+		tighten the possible inputs to just Approve and Cancel.
+	*/
+	switch operation {
```

Comment:
I'd move this up top, and introduce some sort of error condition if the function gets the wrong operation. As an alternative, I'd pass in a bool and set the method based on the bool. This way, you avoid a possible error condition.

Reply:
I actually specifically opted for an enum over a bool to avoid such confusion. I think a boolean is less explicit than case matching. I'm not too familiar with Go so I'm not sure if it will be an exhaustive case match, but it shouldn't be possible to hit any weird state if you require the param to be a `MsigProposeResponse` unless you add a new case (which would be very deliberate). The null case however might be a problem, so it may be worth adding that case if it's a possibility.

Reply:
Gotcha. I'm not necessarily opposed to a boolean, for the reasons you mentioned. The switch just needs to cover all cases, and either add an error condition or throw if passed the wrong method. I prefer the former, even with the slight addition of complexity.